### Import Libraries

In [1]:
import pandas as pd
from collections import OrderedDict
from tqdm.notebook import tqdm
import numpy as np
import os

### Read Weather Data

In [2]:
weather_data_paths = os.listdir("data/weather")
weather_data_paths = [f"data/weather/{path}" for path in weather_data_paths]

In [3]:
# read dataframes
# the data are downloaded in groups ["pr", "tas", "tasmax"] and ["tasmin"] . 
# Because the "tasmin" data is alone, there is a single sheetname called "all"
# Whereas the other excel files contain "pr", "tas", and "tasmax" sheetnames
data_frames = {"pr":[], "tas":[], "tasmax":[], "all":[]}

for weather_data_path in tqdm(weather_data_paths):
    xlsx = xl = pd.ExcelFile(weather_data_path)
    sheet_names = xlsx.sheet_names

    # read data for each sheet in each excel workbook and store the data for the corresponding measurement. Note that
    # "all" == "tasmin"
    for sheet_name in sheet_names:
        df = pd.read_excel(weather_data_path)
        data_frames[sheet_name].append(df)

  0%|          | 0/155 [00:00<?, ?it/s]

It can be seen that one or more datatsets were dataset more that ones. But this will be solved after duplicates are removed from the combined dataset.

In [4]:
# display part of precipitation data
data_frames["pr"][0]

code                 name  1901-07  1902-07  1903-07  1904-07  1905-07  \
0  COG                Congo  1606.94  1629.24  1683.11  1596.47  1626.58   
1  COK  Cook Islands (N.Z.)  2132.50  1871.58  1760.75  1906.54  1747.08   
2  COM              Comoros  2193.48  2189.90  2198.36  2196.18  2194.38   

   1906-07  1907-07  1908-07  ...  2013-07  2014-07  2015-07  2016-07  \
0  1631.87  1636.96  1536.86  ...  1598.65  1521.31  1503.16  1736.68   
1  1805.72  2086.96  2648.38  ...  2060.37  2060.37  2060.37  2060.37   
2  2197.87  1911.27  2083.94  ...  1773.12  1946.99  2182.92  2255.58   

   2017-07  2018-07  2019-07  2020-07  2021-07  2022-07  
0  1744.50  1626.57  1806.57  1734.55  1639.23  1673.04  
1  2060.37  2060.37  2060.37  2060.37  2060.37  2060.37  
2  2093.56  2286.78  1948.26  1980.70  1893.76  2388.39  

[3 rows x 124 columns]

In [5]:
# display part of Average Mean Surface Air Temperature data
data_frames["tas"][0]

code                 name  1901-07  1902-07  1903-07  1904-07  1905-07  \
0  COG                Congo  1606.94  1629.24  1683.11  1596.47  1626.58   
1  COK  Cook Islands (N.Z.)  2132.50  1871.58  1760.75  1906.54  1747.08   
2  COM              Comoros  2193.48  2189.90  2198.36  2196.18  2194.38   

   1906-07  1907-07  1908-07  ...  2013-07  2014-07  2015-07  2016-07  \
0  1631.87  1636.96  1536.86  ...  1598.65  1521.31  1503.16  1736.68   
1  1805.72  2086.96  2648.38  ...  2060.37  2060.37  2060.37  2060.37   
2  2197.87  1911.27  2083.94  ...  1773.12  1946.99  2182.92  2255.58   

   2017-07  2018-07  2019-07  2020-07  2021-07  2022-07  
0  1744.50  1626.57  1806.57  1734.55  1639.23  1673.04  
1  2060.37  2060.37  2060.37  2060.37  2060.37  2060.37  
2  2093.56  2286.78  1948.26  1980.70  1893.76  2388.39  

[3 rows x 124 columns]

In [6]:
# display part of Maximum Mean Surface Air Temperature data
data_frames["tasmax"][0]

code                 name  1901-07  1902-07  1903-07  1904-07  1905-07  \
0  COG                Congo  1606.94  1629.24  1683.11  1596.47  1626.58   
1  COK  Cook Islands (N.Z.)  2132.50  1871.58  1760.75  1906.54  1747.08   
2  COM              Comoros  2193.48  2189.90  2198.36  2196.18  2194.38   

   1906-07  1907-07  1908-07  ...  2013-07  2014-07  2015-07  2016-07  \
0  1631.87  1636.96  1536.86  ...  1598.65  1521.31  1503.16  1736.68   
1  1805.72  2086.96  2648.38  ...  2060.37  2060.37  2060.37  2060.37   
2  2197.87  1911.27  2083.94  ...  1773.12  1946.99  2182.92  2255.58   

   2017-07  2018-07  2019-07  2020-07  2021-07  2022-07  
0  1744.50  1626.57  1806.57  1734.55  1639.23  1673.04  
1  2060.37  2060.37  2060.37  2060.37  2060.37  2060.37  
2  2093.56  2286.78  1948.26  1980.70  1893.76  2388.39  

[3 rows x 124 columns]

In [7]:
# display part of Minimum Mean Surface Air Temperature data
data_frames["all"][0]

code           name  1901-07  1902-07  1903-07  1904-07  1905-07  1906-07  \
0  LBY          Libya    15.21    15.15    15.04    15.07    15.07    15.20   
1  LIE  Liechtenstein     1.75     2.04     2.06     2.66     1.94     1.95   
2  LTU      Lithuania     2.35     0.73     3.26     1.69     2.56     3.16   

   1907-07  1908-07  ...  2013-07  2014-07  2015-07  2016-07  2017-07  \
0    15.01    15.05  ...    16.21    16.12    15.86    16.25    15.82   
1     1.94     1.61  ...     3.60     4.66     4.42     4.06     3.90   
2     1.37     1.93  ...     3.76     3.94     4.52     4.01     4.01   

   2018-07  2019-07  2020-07  2021-07  2022-07  
0    16.32    15.92    15.86    16.28    16.04  
1     4.62     4.35     4.37     3.38     4.73  
2     4.18     4.62     5.11     3.37     3.94  

[3 rows x 124 columns]

### Combine Weather Data For Data Measurements

In [8]:
# arrange pr data
pr_data = {"year":dict(), "country":dict(), "precipitation":dict()}

k= 0
for df in data_frames["pr"]:
    for i in range(df.shape[0]):
        for j in range(2,df.shape[1]):
            pr_value = df.iloc[i,j]
            country = df["name"][i]
            year = df.columns[j].split("-")[0].strip()
            pr_data["precipitation"][k] = pr_value
            pr_data["country"][k] = country
            pr_data["year"][k] = year

            k +=1

pr_data_df = pd.DataFrame.from_dict(pr_data)
pr_data_df.head()

year country  precipitation
0  1901   Congo        1606.94
1  1902   Congo        1629.24
2  1903   Congo        1683.11
3  1904   Congo        1596.47
4  1905   Congo        1626.58

In [9]:
pr_data_df.shape

(28182, 3)

In [10]:
# arrange tas:Average Mean Surface Air Temperature data
tas_data = {"year":dict(), "country":dict(), "average_mean_surface_air_temp":dict()}

k= 0
for df in data_frames["tas"]:
    for i in range(df.shape[0]):
        for j in range(2,df.shape[1]):
            tas_value = df.iloc[i,j]
            country = df["name"][i]
            year = df.columns[j].split("-")[0].strip()
            tas_data["average_mean_surface_air_temp"][k] = tas_value
            tas_data["country"][k] = country
            tas_data["year"][k] = year

            k +=1

tas_data_df = pd.DataFrame.from_dict(tas_data)
tas_data_df.head()

year country  average_mean_surface_air_temp
0  1901   Congo                        1606.94
1  1902   Congo                        1629.24
2  1903   Congo                        1683.11
3  1904   Congo                        1596.47
4  1905   Congo                        1626.58

In [11]:
tas_data_df.shape

(28182, 3)

In [12]:
# arrange tasmax:Average Maximum Surface Air Temperature data
tasmax_data = {"year":dict(), "country":dict(), "average_maximum_surface_air_temp":dict()}

k= 0
for df in data_frames["tasmax"]:
    for i in range(df.shape[0]):
        for j in range(2,df.shape[1]):
            tasmax_value = df.iloc[i,j]
            country = df["name"][i]
            year = df.columns[j].split("-")[0].strip()
            tasmax_data["average_maximum_surface_air_temp"][k] = tasmax_value
            tasmax_data["country"][k] = country
            tasmax_data["year"][k] = year

            k +=1

tasmax_data_df = pd.DataFrame.from_dict(tasmax_data)
tasmax_data_df.head()

year country  average_maximum_surface_air_temp
0  1901   Congo                           1606.94
1  1902   Congo                           1629.24
2  1903   Congo                           1683.11
3  1904   Congo                           1596.47
4  1905   Congo                           1626.58

In [13]:
tasmax_data_df.shape

(28182, 3)

In [14]:
# arrange tasmin:Average Minimum Surface Air Temperature data
tasmin_data = {"year":dict(), "country":dict(), "average_minimum_surface_air_temp":dict()}

k= 0
for df in data_frames["all"]:
    for i in range(df.shape[0]):
        for j in range(2,df.shape[1]):
            tasmin_value = df.iloc[i,j]
            country = df["name"][i]
            year = df.columns[j].split("-")[0].strip()
            tasmin_data["average_minimum_surface_air_temp"][k] = tasmin_value
            tasmin_data["country"][k] = country
            tasmin_data["year"][k] = year

            k +=1

tasmin_data_df = pd.DataFrame.from_dict(tasmin_data)
tasmin_data_df.head()

year country  average_minimum_surface_air_temp
0  1901   Libya                             15.21
1  1902   Libya                             15.15
2  1903   Libya                             15.04
3  1904   Libya                             15.07
4  1905   Libya                             15.07

In [15]:
tasmin_data_df.shape

(28426, 3)

In [16]:
## Let's examine if all countries are well represented
countries_in_pr = set(pr_data_df.country.unique())
countries_in_tas = set(tas_data_df.country.unique())
countries_in_tasmax = set(tasmax_data_df.country.unique())
countries_in_tasmin = set(tasmin_data_df.country.unique())

common_countries = countries_in_pr.intersection(countries_in_tas, countries_in_tasmax, countries_in_tasmin)
all_countries = countries_in_pr.union(countries_in_tas, countries_in_tasmax, countries_in_tasmin)
num_common_countries = len(common_countries)

print(f"Number of countries in pr dataset is: {len(countries_in_pr)}")
print(f"Number of countries in tas dataset is: {len(countries_in_tas)}")
print(f"Number of countries in tasmax dataset is: {len(countries_in_tasmax)}")
print(f"Number of countries in tasmin dataset is: {len(countries_in_tasmin)}")
print(f"Number of countries common to all datasets : {num_common_countries}")
print(f"Number of included countries: {len(all_countries)}")

Number of countries in pr dataset is: 225
Number of countries in tas dataset is: 225
Number of countries in tasmax dataset is: 225
Number of countries in tasmin dataset is: 202
Number of countries common to all datasets : 200
Number of included countries: 227


In [17]:
# get countries that donot have all measurements:
missing_some_measurement_countries = {country for country in all_countries if country not in common_countries}
missing_some_measurement_countries

{'American Samoa (U.S.)',
 'Anguilla (U.K.)',
 'Aruba (Neth.)',
 'Bermuda (U.K.)',
 'Bouvet Island (Nor.)',
 'British Indian Ocean Territory (U.K.)',
 'British Virgin Islands (U.K.)',
 'Cayman Islands (U.K.)',
 'Christmas Island (Aus.)',
 'Cocos (Keeling) Islands (Aus.)',
 'Cook Islands (N.Z.)',
 'Curaçao (Neth.)',
 'Faroe Islands (Den.)',
 'French Polynesia (Fr.)',
 'French Southern and Antarctic Lands (Fr.)',
 'Gibraltar (U.K.)',
 'Greenland (Den.)',
 'Guam (U.S.)',
 'Isle of Man (U.K.)',
 'Jersey (U.K.)',
 'Montserrat (U.K.)',
 'New Caledonia (Fr.)',
 'Puerto Rico (U.S.)',
 'Saint Helena, Ascension and Tristan da Cunha (U.K.)',
 'Saint-Barthélemy (Fr.)',
 'Saint-Martin (Fr.)',
 'Saint-Pierre-et-Miquelon (Fr.)'}

Oh this makes sense. At first the data was downloaded with no regards to whether it is for a country or a state. But after haven noticed, only data for countries was downloaded. So, we will check in the common countries as well if there is any country name that is a state and remove them.


In [18]:
# state names contains brackets
for country_name in common_countries:
    if "(" in country_name:
        print(country_name)

Tokelau (N.Z.)
Niue (N.Z.)


In [19]:
remove_list = []
for country in all_countries:
    if "(" in country:
        remove_list.append(country)
# clean common countries
common_countries = [country for country in common_countries if country not in remove_list]
len(common_countries)

198

Now we will remove all states from all datasets.

In [20]:
# filter pr data
pr_data_df = pr_data_df[pr_data_df['country'].isin(common_countries)].drop_duplicates().reset_index(drop=True)

# filter tas data
tas_data_df = tas_data_df[tas_data_df['country'].isin(common_countries)].drop_duplicates().reset_index(drop=True)

# filter tasmax data
tasmax_data_df = tasmax_data_df[tasmax_data_df['country'].isin(common_countries)].drop_duplicates().reset_index(drop=True)

# filter tasmin data
tasmin_data_df = tasmin_data_df[tasmin_data_df['country'].isin(common_countries)].drop_duplicates().reset_index(drop=True)

In [21]:
## Let's examine if all countries are well represented again
countries_in_pr = set(pr_data_df.country.unique())
countries_in_tas = set(tas_data_df.country.unique())
countries_in_tasmax = set(tasmax_data_df.country.unique())
countries_in_tasmin = set(tasmin_data_df.country.unique())

common_countries = countries_in_pr.intersection(countries_in_tas, countries_in_tasmax, countries_in_tasmin)
all_countries = countries_in_pr.union(countries_in_tas, countries_in_tasmax, countries_in_tasmin)
num_common_countries = len(common_countries)

print(f"Number of countries in pr dataset is: {len(countries_in_pr)}")
print(f"Number of countries in tas dataset is: {len(countries_in_tas)}")
print(f"Number of countries in tasmax dataset is: {len(countries_in_tasmax)}")
print(f"Number of countries in tasmin dataset is: {len(countries_in_tasmin)}")
print(f"Number of countries common to all datasets : {num_common_countries}")
print(f"Number of included countries: {len(all_countries)}")

Number of countries in pr dataset is: 198
Number of countries in tas dataset is: 198
Number of countries in tasmax dataset is: 198
Number of countries in tasmin dataset is: 198
Number of countries common to all datasets : 198
Number of included countries: 198


Perfect! Now there are 198 countries.

## Combine All Measurements
In order to ensure that measurements are matched correctly. We will match measurements according to 2 columns:
* country and
* year.

In [22]:
combined_weather_data = {"year":dict(), 
                 "country":dict(), 
                 "precipitation":dict(),
                 "average_mean_surface_air_temp":dict(),
                 "average_maximum_surface_air_temp":dict(),
                 "average_minimum_surface_air_temp":dict()
                }

# ieterating over pr data and finding a match for tas, tasmax and tasmin data
for i in tqdm(range(len(pr_data_df))):
    year = pr_data_df["year"][i]
    country = pr_data_df["country"][i]
    precipitation = pr_data_df["precipitation"][i]

    # search for tas measurement where country and year are same as for pr measurement.
    mean_temp_index = np.where((tas_data_df["year"] == year) & (tas_data_df["country"]==country))[0][0]
    mean_temp = tas_data_df["average_mean_surface_air_temp"][mean_temp_index]

    # search for tasmax measurement where country and year are same as for pr measurement.
    max_temp_index = np.where((tasmax_data_df["year"] == year) & (tasmax_data_df["country"]==country))[0][0]
    max_temp = tasmax_data_df["average_maximum_surface_air_temp"][max_temp_index]

    # search for tasmin measurement where country and year are same as for pr measurement.
    min_temp_index = np.where((tasmin_data_df["year"] == year) & (tasmin_data_df["country"]==country))[0][0]
    min_temp = tasmin_data_df["average_minimum_surface_air_temp"][min_temp_index]

    # populate combined date
    combined_weather_data["year"][i] = year
    combined_weather_data["country"][i] = country
    combined_weather_data["precipitation"][i] = precipitation
    combined_weather_data["average_mean_surface_air_temp"][i] = mean_temp
    combined_weather_data["average_maximum_surface_air_temp"][i] = max_temp
    combined_weather_data["average_minimum_surface_air_temp"][i] = min_temp

combined_weather_data_df = pd.DataFrame.from_dict(combined_weather_data)
combined_weather_data_df.head()

  0%|          | 0/24156 [00:00<?, ?it/s]

year country  precipitation  average_mean_surface_air_temp  \
0  1901   Congo        1606.94                        1606.94   
1  1902   Congo        1629.24                        1629.24   
2  1903   Congo        1683.11                        1683.11   
3  1904   Congo        1596.47                        1596.47   
4  1905   Congo        1626.58                        1626.58   

   average_maximum_surface_air_temp  average_minimum_surface_air_temp  
0                           1606.94                             19.72  
1                           1629.24                             19.83  
2                           1683.11                             19.92  
3                           1596.47                             19.71  
4                           1626.58                             19.98

In [23]:
combined_weather_data_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 24156 entries, 0 to 24155
Data columns (total 6 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   year                              24156 non-null  object 
 1   country                           24156 non-null  object 
 2   precipitation                     24156 non-null  float64
 3   average_mean_surface_air_temp     24156 non-null  float64
 4   average_maximum_surface_air_temp  24156 non-null  float64
 5   average_minimum_surface_air_temp  24156 non-null  float64
dtypes: float64(4), object(2)
memory usage: 1.3+ MB


Apparently there are no missing values in the weather dataset

### Read Malaria Incidence Data

In [24]:
malaria_data = pd.read_csv("data/malaria-incidence.csv")
malaria_data.head()

IND_UUID       IND_NAME               IND_CODE  DIM_GEO_CODE_M49  \
0  442CEA8  Malaria cases  MALARIA_EST_INCIDENCE                 4   
1  442CEA8  Malaria cases  MALARIA_EST_INCIDENCE               953   
2  442CEA8  Malaria cases  MALARIA_EST_INCIDENCE                24   
3  442CEA8  Malaria cases  MALARIA_EST_INCIDENCE               954   
4  442CEA8  Malaria cases  MALARIA_EST_INCIDENCE                32   

  GEO_NAME_SHORT  DIM_TIME DIM_TIME_TYPE DIM_VALUE_TYPE  VALUE_NUMERIC  \
0    Afghanistan      2000          YEAR  RATE_PER_1000       87.15783   
1         Africa      2000          YEAR  RATE_PER_1000      372.59359   
2         Angola      2000          YEAR  RATE_PER_1000      325.70361   
3       Americas      2000          YEAR  RATE_PER_1000       14.12108   
4      Argentina      2000          YEAR  RATE_PER_1000        2.37385   

   VALUE_NUMERIC_LOWER  VALUE_NUMERIC_UPPER DIM_PUBLISH_STATE_CODE  
0             55.97768            134.25330              PUBLISHED  
1            343.83142            404.60421              PUBLISHED  
2            197.40626            506.60785              PUBLISHED  
3             12.76521             15.58999              PUBLISHED  
4              2.37385              2.37385              PUBLISHED

In [25]:
malaria_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2478 entries, 0 to 2477
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   IND_UUID                2478 non-null   object 
 1   IND_NAME                2478 non-null   object 
 2   IND_CODE                2478 non-null   object 
 3   DIM_GEO_CODE_M49        2478 non-null   int64  
 4   GEO_NAME_SHORT          2478 non-null   object 
 5   DIM_TIME                2478 non-null   int64  
 6   DIM_TIME_TYPE           2478 non-null   object 
 7   DIM_VALUE_TYPE          2478 non-null   object 
 8   VALUE_NUMERIC           2478 non-null   float64
 9   VALUE_NUMERIC_LOWER     2477 non-null   float64
 10  VALUE_NUMERIC_UPPER     2477 non-null   float64
 11  DIM_PUBLISH_STATE_CODE  2478 non-null   object 
dtypes: float64(3), int64(2), object(7)
memory usage: 232.4+ KB


In [26]:
# Some countries are present in both datasets but do not match in names.
# After a manual inspection, their names are modified to match
combined_weather_data_df = combined_weather_data_df.replace(["The Gambia", 
                                                             "São Tomé and Príncipe",
                                                             "Democratic Republic of Congo",
                                                             "D. P. R. of Korea",
                                                             "Tanzania",
                                                             "Islamic Republic of Iran"
                                                            ], 
                                                            ["Gambia", 
                                                             "Sao Tome and Principe",
                                                             'Democratic Republic of the Congo',
                                                             'Republic of Korea',
                                                             'United Republic of Tanzania',
                                                             'Iran (Islamic Republic of)',
                                                             
                                                            ])

malaria_data_countries = malaria_data["GEO_NAME_SHORT"].unique()
weather_data_countries = combined_weather_data_df["country"].unique()

In [27]:
# how many west african countries have matched malaria incidence data?
matched_countries = []
mismatched_countries = []
for country in weather_data_countries:
    if country in malaria_data_countries:
        matched_countries.append(country)
    else:
        mismatched_countries.append(country)

print("Number of matched countries: ",len(matched_countries))

Number of matched countries:  101


Great! Apparently, though many countries donot have malaria data, significant number of countries are matched.

Now let's filter the weather dataset to include only countries that are matched with the malaria dataset.

In [28]:
combined_weather_data_df = combined_weather_data_df[combined_weather_data_df['country'].isin(matched_countries)]
combined_weather_data_df = combined_weather_data_df.reset_index(drop =True)
combined_weather_data_df.shape

(12444, 6)

Now there are 12444 weather observations spanning from 1901 to 2022. However, the malaria dataset spans from from the year 2000 to 2021. Therefore, the number of observations in the combined dataset will drop significantly.

### Combine both datasets
To combine both datasets, we only need to obtain the malaria incidence values for each year and each country from the malaria dataset and match it to the corresponding country and year in the weather dataset.

In [29]:
combined_weather_data_df["year"] = combined_weather_data_df["year"].astype("int")
malaria_data["DIM_TIME"] = malaria_data["DIM_TIME"].astype("int")
malaria_years = [year for year in malaria_data["DIM_TIME"].unique()]
n = combined_weather_data_df.shape[0]
m = malaria_data.shape[0]

weather_match_indices = {f'{combined_weather_data_df["country"][i]}_and_{combined_weather_data_df["year"][i]}':i for i in range(n)}
incidence_matches = [f'{malaria_data["GEO_NAME_SHORT"][i]}_and_{malaria_data["DIM_TIME"][i]}' for i in range(m)]

malaria_incidence_values = {"incidence":{}}
k=0
others = []
for i in tqdm(range(combined_weather_data_df.shape[0])):
    year = combined_weather_data_df["year"][i]
    country = combined_weather_data_df["country"][i]
    if f"{country}_and_{year}" in incidence_matches:
        k +=1

        index = np.where((malaria_data["DIM_TIME"] == year) & (malaria_data["GEO_NAME_SHORT"]==country))[0][0]
        incidence = malaria_data["VALUE_NUMERIC"][index]
            
    else:
        incidence = None
    malaria_incidence_values["incidence"][i] = incidence

incidence_df = pd.DataFrame.from_dict(malaria_incidence_values)
incidence_df.tail()

  0%|          | 0/12444 [00:00<?, ?it/s]

incidence
12439    0.27837
12440    0.18801
12441    0.09468
12442    0.11231
12443        NaN

In [30]:
# combine all data
data = pd.concat([combined_weather_data_df, pd.DataFrame.from_dict(malaria_incidence_values)], axis = 1)
# remove None values
data = data[data["incidence"].notna()].reset_index(drop = True)
data.head()

year country  precipitation  average_mean_surface_air_temp  \
0  2000   Congo        1644.79                        1644.79   
1  2001   Congo        1516.01                        1516.01   
2  2002   Congo        1717.96                        1717.96   
3  2003   Congo        1573.98                        1573.98   
4  2004   Congo        1507.59                        1507.59   

   average_maximum_surface_air_temp  average_minimum_surface_air_temp  \
0                           1644.79                             19.99   
1                           1516.01                             20.19   
2                           1717.96                             20.26   
3                           1573.98                             20.23   
4                           1507.59                             20.33   

   incidence  
0  353.41557  
1  350.93625  
2  321.67402  
3  319.21132  
4  317.81208

In [31]:
data.shape

(2178, 7)

The combined dataset is made up of 2178 entries which are enough for modeling.

### Adding Longitude and Latitude Information

In [32]:
# load location data
locations_df = pd.read_csv("data/countries.csv")
locations_df.head()

country   latitude  longitude                  name
0      AD  42.546245   1.601554               Andorra
1      AE  23.424076  53.847818  United Arab Emirates
2      AF  33.939110  67.709953           Afghanistan
3      AG  17.060816 -61.796428   Antigua and Barbuda
4      AI  18.220554 -63.068615              Anguilla

In [33]:
# Make sure the countries match
# Some countries are present in both datasets but do not match in names.
# After a manual inspection, their names are modified to match
locations_df = locations_df.replace([
                                    'Congo [Republic]',
                                    'Congo [DRC]',
                                    'Turkey',
                                    'São Tomé and Príncipe',
                                    'Iran',
                                    'Cape Verde',
                                    'Myanmar [Burma]',
                                    'Syria',
                                    'Laos'
                                    ], 
                                    [
                                    "Congo",
                                    "Democratic Republic of the Congo",
                                    "Türkiye",
                                    "Sao Tome and Principe",
                                    "Iran (Islamic Republic of)",
                                    "Cabo Verde",
                                    "Myanmar",
                                    "Syrian Arab Republic",
                                    "Lao People's Democratic Republic"

                                    ])
location_data_countries = locations_df.name.unique()
combined_data_countries = data.country.unique()
mismatched = []
for country in combined_data_countries:
    if country not in location_data_countries:
        mismatched.append(country)
        print(country)

Republic of Korea
United Republic of Tanzania
South Sudan
Eswatini


Four countries do not match. Therefore they must be removed from the the dataset.

In [34]:
# Remove mismatched countries and add longitude and lattitude data
data = data[-data.country.isin(mismatched)].reset_index(drop = True)
data["longitude"] = data["country"].apply(lambda x: locations_df[locations_df["name"] == x]["longitude"].values[0] if x in locations_df["name"].values else None)
data["latitude"] = data["country"].apply(lambda x: locations_df[locations_df["name"] == x]["latitude"].values[0] if x in locations_df["name"].values else None)


In [35]:
data.head()

year country  precipitation  average_mean_surface_air_temp  \
0  2000   Congo        1644.79                        1644.79   
1  2001   Congo        1516.01                        1516.01   
2  2002   Congo        1717.96                        1717.96   
3  2003   Congo        1573.98                        1573.98   
4  2004   Congo        1507.59                        1507.59   

   average_maximum_surface_air_temp  average_minimum_surface_air_temp  \
0                           1644.79                             19.99   
1                           1516.01                             20.19   
2                           1717.96                             20.26   
3                           1573.98                             20.23   
4                           1507.59                             20.33   

   incidence  longitude  latitude  
0  353.41557  15.827659 -0.228021  
1  350.93625  15.827659 -0.228021  
2  321.67402  15.827659 -0.228021  
3  319.21132  15.827659 -0.228021  
4  317.81208  15.827659 -0.228021

In [36]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2068 entries, 0 to 2067
Data columns (total 9 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   year                              2068 non-null   int64  
 1   country                           2068 non-null   object 
 2   precipitation                     2068 non-null   float64
 3   average_mean_surface_air_temp     2068 non-null   float64
 4   average_maximum_surface_air_temp  2068 non-null   float64
 5   average_minimum_surface_air_temp  2068 non-null   float64
 6   incidence                         2068 non-null   float64
 7   longitude                         2068 non-null   float64
 8   latitude                          2068 non-null   float64
dtypes: float64(7), int64(1), object(1)
memory usage: 145.5+ KB


Perfect! Our final dataset contains 2068 observations.

In [37]:
# save data
data.to_csv("data/processed_data.csv", index = False)